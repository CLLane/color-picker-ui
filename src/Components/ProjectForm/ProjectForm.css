.project-form__container {
  width: 100%;
  display: flex;
  justify-content: center;
  align-items: center;
}

.project-name__input, .palette-name__input, .project-name__dropdown {
  margin-right: 10px;
  margin-bottom: 10px;
  margin-top: 10px;
  background: rgb(245, 245, 245, .9);
  border: 2px solid grey;
}

.project__form {
  background: rgb(245, 245, 245, .8);
  border-radius: 8px;
  padding: 5px;
  border: solid 4px darkgrey;
}