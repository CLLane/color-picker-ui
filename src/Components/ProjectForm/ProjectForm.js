import React, { Component } from 'react';
import './ProjectForm.css';

export class ProjectForm extends Component {
  constructor() {
    super();
    this.state = {
      palette: null,
      isExistingProject: false,
      projectToSave: '',
      paletteToSave: '',
      selectValue: 'new-project'
    }
  };

  selectProject = (value) => {
    const { projects } = this.props;
    let currentProject = projects.find(project => project.name === value);
    if ( currentProject ) {
      this.setState({ projectToSave: currentProject, isExistingProject: true, selectValue:value })
    } else {
      this.setState({ projectToSave: "", isExistingProject: false, selectValue: 'new-project'})
    }
  }

  handleNameChange = (e) => {
    this.setState({ [e.target.name]: e.target.value })
  }

  handleSubmit = async (e) => { 
    e.preventDefault();
    const { projectToSave, paletteToSave } = this.state;
    const { handleSubmission } = this.props;
    await handleSubmission(projectToSave, paletteToSave);
    const nameValue = projectToSave.name ? projectToSave.name : projectToSave;
    this.setState({ selectValue: nameValue, isExistingProject: true, paletteToSave: "" })
  }

  render() {
    const { projects } = this.props;
    const { isExistingProject, projectToSave, paletteToSave, selectValue } = this.state;
    const projectPlaceholderText = isExistingProject ? projectToSave.name : 'Enter Project Name';
    const projectNameValue = projectToSave.name ? projectToSave.name : projectToSave;
    const options = projects.map((project, index) => {
      return <option key={index} value={project.name}>{project.name}</option>
    })
    return (
      <form>
        <select value={selectValue} onChange={(e) => this.selectProject(e.target.value)}>
          { options }
          <option value='new-project'>Create New Project</option>
        </select>
        <input type='text' disabled={isExistingProject} name='projectToSave' placeholder={ projectPlaceholderText } onChange={this.handleNameChange} value={projectNameValue}></input>
        <input type='text' placeholder='Enter Palette Name' name='paletteToSave' onChange={this.handleNameChange} value={paletteToSave}></input>
        <button onClick={this.handleSubmit}>Save</button>
      </form>
    )
  }
}