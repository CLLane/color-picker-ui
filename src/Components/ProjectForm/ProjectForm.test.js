import React from 'react';
import { shallow } from 'enzyme';
import ProjectForm from './ProjectForm';
import './ProjectForm.css';

describe('LoginForm', () => {
  let wrapper;
  let mockError;
  let mockProject;
  let mockProjects;
  let mockUpdateCurrentProject;
  let mockHandleSubmission;
  beforeEach(() => {
    mockProjects = [{
      id: 12345,
      name: 'Cool Project',
      user_id: 1
    }];
    mockProject = {
      id: 12345,
      name: 'Cool Project',
      user_id: 1
    };
    mockError = '';
    mockUpdateCurrentProject = jest.fn();
    mockHandleSubmission = jest.fn();
    wrapper = shallow(<ProjectForm
      error={mockError}
      currentProject={mockProject} 
      projects={mockProjects}
      updateCurrentProject={mockUpdateCurrentProject} 
      handleSubmission={mockHandleSubmission} 
      />);
  });

  it('should match the snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should update state when handle name change is called', () => {
    const mockEvent = {
      target: {
        name: 'projectName', 
        value: 'cool project'
      }
    };
    wrapper.instance().handleNameChange(mockEvent);
    const expectedProject = 'cool project';
    expect(wrapper.state('projectName')).toEqual(expectedProject);
  });

  it('should set state to new-project when select project is called, and set the current project to null on App', () => {
    wrapper.setState({select: 'bob'});
    wrapper.instance().selectProject('new-project');
    const expectedState = 'new-project';
    expect(mockUpdateCurrentProject).toHaveBeenCalledWith(null);
    expect(wrapper.state('select')).toEqual(expectedState);
  });

  it('if a project exists, it should set the project name to state', () => {
    wrapper.setState({select: 'cool project', paletteName:'cool palette'});
    const mockEvent = {
      preventDefault: jest.fn()
    };
    wrapper.instance().handleSubmit(mockEvent);
    expect(mockHandleSubmission).toHaveBeenCalledWith('cool project', 'cool palette')
  });

});