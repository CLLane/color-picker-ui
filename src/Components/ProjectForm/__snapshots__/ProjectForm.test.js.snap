// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginForm should match the snapshot 1`] = `
<div
  className="project-form__container"
>
  <form
    className="project__form"
  >
    <select
      className="project-name__dropdown"
      onChange={[Function]}
      value="new-project"
    >
      <option
        key="0"
        value="Cool Project"
      >
        Cool Project
      </option>
      <option
        value="new-project"
      >
        Create New Project
      </option>
    </select>
    <input
      className="project-name__input"
      disabled={true}
      maxlength="25"
      name="projectName"
      onChange={[Function]}
      placeholder="Cool Project"
      type="text"
      value=""
    />
    <input
      className="palette-name__input"
      maxlength="13"
      name="paletteName"
      onChange={[Function]}
      placeholder="Enter Palette Name"
      type="text"
      value=""
    />
    <img
      alt="disable save palette"
      src="disableNewPalette.svg"
    />
  </form>
</div>
`;
